plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -797,35 +822,97 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
+               if( my->chain_id ) {
+                  ilog("REMOVE chain id: ${ci}, genesis chain id: ${gci}", ("ci", *my->chain_id)("gci",genesis_file.compute_chain_id()));
+                  EOS_ASSERT( *my->chain_id == genesis_file.compute_chain_id(), plugin_config_exception,
+                              "Genesis state's chain id provided via command line arguments does not match the chain id in the ${source}. "
+                              "It is not necessary to provide a genesis state argument when loading a snapshot.", ("source", my->chain_id_source)
+                  );
+               }
+               else {
+                  my->chain_id = genesis_file.compute_chain_id();
+                  my->chain_id_source = "genesis json";
+               }
+            }
          }
+
+         if( options.count( "chain-id" ) ) {
+            auto ci = options.at( "chain-id" );
+            ilog("chain-id");
+            ilog("chain-id passed ${ci}",("ci", ci.as<string>()));
+            auto get_chain_id = [chain_id=ci.as<string>()]() {
+               try {
+                  ilog("chain-id: ${ci}",("ci", chain_id));
+                  return fc::json::from_string(chain_id).as<chain_id_type>();
+               } catch ( fc::exception& e ) {
+                  elog("Malformed chain id");
+                  throw e;
+               }};
+            const chain_id_type passed_chain_id = get_chain_id();
+            if( my->chain_id ) {
+               ilog("REMOVE chain id: ${ci}, passed_chain_id: ${gci}", ("ci", *my->chain_id)("gci",passed_chain_id));
+               EOS_ASSERT( *my->chain_id == passed_chain_id, plugin_config_exception,
+                           "chain id provided via command line arguments does not match the existing chain id in the ${source}. "
+                           "It is not necessary to provide a chain id argument when loading a snapshot.", ("source", my->chain_id_source)
+               );
+            }
+            else {
+               my->chain_id = passed_chain_id;
+               my->chain_id_source = "chain id command line parameter";
+            }
+         }
+
          auto shared_mem_path = my->chain_config->state_dir / "shared_memory.bin";
          EOS_ASSERT( !fc::exists(shared_mem_path),
                  plugin_config_exception,
                  "Snapshot can only be used to initialize an empty database." );
 
          if( fc::is_regular_file( my->blocks_dir / "blocks.log" )) {
-            auto log_genesis = block_log::extract_genesis_state(my->blocks_dir);
-            EOS_ASSERT( log_genesis.compute_chain_id() == my->chain_config->genesis.compute_chain_id(),
-                    plugin_config_exception,
-                    "Genesis information in blocks.log does not match genesis information in the snapshot");
+            const auto log_chain_id = block_log::extract_chain_id(my->blocks_dir);
+            if( my->chain_id ) {
+               EOS_ASSERT( log_chain_id == *my->chain_id,
```

Comment:
Assert that `log_chain_id == snapshot_chain_id`, then store either one into a `optional<chain_id_type>` member variable in `chain_plugin_impl` called `expected_chain_id`.